src/main/java/tech/stackable/spark/operator/abstractcontroller/AbstractCrdController.java

```
@@ -0,0 +1,318 @@
+      namespace = "default";
+      LOGGER.trace("No namespace found via config, assuming " + namespace);
+    }
+  }
+
+  /**
+   * load crd context, init informers and crdClient, write crd and register event handlers
+   */
+  @SuppressWarnings("unchecked")
+  public void init() {
+    crdMetadata = loadYaml(crdPath);
+
+    informerFactory = client.informers();
+
+    crdContext = getCrdContext(crdMetadata);
+
+    crdSharedIndexInformer = informerFactory.sharedIndexInformerForCustomResource(
+      crdContext,
+      (Class<Crd>) ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[0],
+      (Class<CrdList>) ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[1],
+      resyncCycle
+    );
+
+    crdLister = new Lister<>(crdSharedIndexInformer.getIndexer(), namespace);
+
+    crdClient = client.customResources(
+      crdContext,
+      (Class<Crd>) ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[0],
+      (Class<CrdList>) ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[1],
+      (Class<CrdDoneable>) ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[2]
+    );
+
+    // initialize crd -> should be one for each controller
+    writeCrd(namespace, crdMetadata);
+
+    // register events handler
+    registerCrdEventHandler();
+  }
+
+  /**
+   * Build crd context for the sharedIndexInformer or other operations from the crd yaml spec
+   *
+   * @return crd context
+   */
+  protected static CustomResourceDefinitionContext getCrdContext(List<HasMetadata> metadata) {
+
+    CustomResourceDefinitionContext.Builder builder = new CustomResourceDefinitionContext.Builder();
+    // check metadata
+    if (metadata.isEmpty()) {
+      LOGGER.warn("No metadata available - return null");
+      return null;
+    }
+    if (metadata.size() > 1) {
+      LOGGER.warn("multiple crds available ... using first one");
```

I know this is only used with hardcoded values for now and CRDs we ship ourselves but if you move this to a library later I'm in favor of erroring out in these kinds of cases.

Maybe instead of accepting a list have this just accept a single `CustomResourceDefinition` parameter so the caller has to decide how she wants to handle this.